### DISCUSSION TITLE
Route Handler + cache + calling from server component

### DISCUSSION THREAD

**Discussion by mohidmakhdoomi** [vercel/next.js#68179] [Help] [answered]
### Summary

As part of an App Router application I have a route handler `/app/graph/route.ts` with a simple GET method

```
import { read } from '@/lib/neo4j'; // this is my library for connecting to neo4j db

export async function GET() {
  // I don't touch any request parameters or use any Dynamic Functions
  const data = await read(...);
  return Response.json(data);
}
```

I have a separate page `/app/page.tsx` where I would like to use `/app/graph/route.ts` 

```
import {GET} from "./graph/route"

export default async function Index() {
  const fetchData = await GET();
  const resultData = JSON.stringify(await fetchData.json());
  ...
  return ...
}
```

Having trouble with the following:

1. Set settings for `/app/graph/route.ts` such that it is cached for say 15 seconds and then next time `/graph` endpoint is visited the data is re-fetched
2. Within `/app/page.tsx` use the GET function from `/app/graph/route.ts` in such a way that the result of GET is cached for say 60 seconds

For 2. I was using `fetch` originally but based on this video I believe using `fetch` wouldn't be ideal. [Vercel: 10 common mistakes with the Next.js App Router ](https://www.youtube.com/watch?v=RBM03RihZVs&t=69s)

**Comment by Juneezee** [comment]
You don't need the Route Handler unless you are planning to expose it as API to the external.

Instead, you can fetch from your Neo4j database directory in your `page.tsx`. `page.tsx` must be a server component.

```tsx
import { read } from '@/lib/neo4j';

export default async function Index() {
  const data = await read(...);
  ...
  return ...
}
```

If you would like to cache the response from Neo4j database, you can use [`unstable_cache` API](https://nextjs.org/docs/app/api-reference/functions/unstable_cache).

See more [Fetching data on the Server with third-party libraries](https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#fetching-data-on-the-server-with-third-party-libraries).

**Reply by mohidmakhdoomi** [reply, marked as answer]
@Juneezee thanks for the reply! yup I did want the /graph endpoint to stay as an external API. 

But i think i solved the problem, 

for 1. i just had to use Route Segment option `revalidate` like so

```
import { read } from '@/lib/neo4j'; // this is my library for connecting to neo4j db

export const revalidate = 15

export async function GET() {
  // I don't touch any request parameters or use any Dynamic Functions
  const data = await read(...);
  return Response.json(data);
}

```

and for 2. I just used unstable_cache in my `/app/page.tsx` like so

```
import {GET} from "./graph/route"
import {unstable_cache} from "next/cache";

const getCachedGet = unstable_cache(
    async () => (await GET().then(async response => await response.json())),
    [],
    {revalidate: 60},
);

export default async function Index() {
  const resultData = JSON.stringify(await getCachedGet());
  ...
  return ...
}

```

Also I had tried both of the above things earlier but the big difference was that this time i used `npm run start` after doing a build... didn't realize the `npm run dev` did not handle caching the same as production
